### PR TITLE
Support prerequisite transactions in dynamic tables write

### PR DESCRIPTION
**Pull request by gritukan**
We are building a service that uses dynamic tables as a metadata storage. Cypress is used for a leader election and prerequisite transactions are used to prevent races, however prerequisite transactions are not supported for dynamic table operations. This PR fixes that.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: yt/yt/server/master/cell_server/tamed_cell_manager.cpp [hunk 1]
```diff
@@ -950,15 +952,39 @@ class TTamedCellManager
...
-    void OnCellDecommissionedOnNode(TCellBase* cell)
+    void OnCellDecommissionedOnNode(TCellBase* cell, bool forceRemove)
     {
         if (cell->IsDecommissionCompleted()) {
             return;
         }
 
+        const auto& transactionManager = Bootstrap_->GetTransactionManager();
+
+        auto leaseTransactionIds = cell->LeaseTransactionIds();
```
Comment (zlobober) [inline yt/yt/server/master/cell_server/tamed_cell_manager.cpp:966]: const auto&?

File: yt/yt/server/node/tablet_node/tablet_cell_write_manager.cpp [hunk 2]
```diff
@@ -624,12 +669,20 @@ class TTabletCellWriteManager
 
                 AddPersistentAffectedTablet(transaction, tablet);
 
+                auto leaseManager = GetLeaseManager();
+                for (auto prerequisiteTransactionId : prerequisiteTransactionIds) {
+                    auto* lease = leaseManager->GetLease(prerequisiteTransactionId);
+                    transaction->PersistentLeaseGuards().push_back(lease->GetPersistentLeaseGuard(/*force*/ true));
+                }
```
Comment (savrus) [inline yt/yt/server/node/tablet_node/tablet_cell_write_manager.cpp:676]: Копипаста с кодом выше. Может сделаем хелпер?

File: yt/yt/ytlib/api/native/client_internal_impl.cpp [hunk 3]
```diff
@@ -148,6 +153,86 @@ void TClient::DoUnlockHunkStore(
         options);
 }
 
+void TClient::DoIssueLease(
+    TCellId cellId,
+    TObjectId leaseId,
+    const TIssueLeaseOptions& options)
+{
+    const auto& cellDirectorySynchronizer = Connection_->GetCellDirectorySynchronizer();
+    WaitFor(cellDirectorySynchronizer->Sync())
```
Comment (savrus) [inline yt/yt/ytlib/api/native/client_internal_impl.cpp:162]: Давай сделаем хелпер GetSyncCellChannelOrThrow, чтобы запускать синк, только если cell directory ещё не знает про целл

File: yt/yt/tests/integration/dynamic_tables/test_dynamic_tables_leases.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+    def _check_lease(self, cell_id, state, prc, trc, lease_id=LEASE_ID):
+        leases = self._get_leases(cell_id)
+        assert len(leases) == 1
+        assert lease_id in leases
+        lease = leases[lease_id]
+        assert lease["state"] == state
+        assert lease["persistent_ref_counter"] == prc
+        assert lease["transient_ref_counter"] == trc
+
+    def _restart_cell(self):
```
Comment (savrus) [inline yt/yt/tests/integration/dynamic_tables/test_dynamic_tables_leases.py:38]: Есть такая штука: https://github.com/ytsaurus/ytsaurus/blob/4c3689e0117cfb69397144e473d393b420dd264d/yt/yt/tests/library/yt_dynamic_tables_base.py#L34

**Review by savrus — CHANGES_REQUESTED** [review]

File: yt/yt/server/lib/lease_server/lease_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,970 @@
...
+        lease->SetPersistentRefCounter(persistentRefCounter);
+
+        YT_LOG_DEBUG(
+            "Lease unreferenced persistently "
+            "(LeaseId: %v, PersistentRefCounter: %v -> %v)",
+            lease->GetId(),
+            persistentRefCounter + 1,
+            persistentRefCounter);
+
+        MaybeRemoveLease(lease);
```
[the remark this answers is not in the document]
Comment (gritukan) [reply yt/yt/server/lib/lease_server/lease_manager.cpp:750]: Good observation!

**Comment by savrus** [comment]
Shipped!

**Comment by savrus** [comment]
Shipped!

**Pull request closed without merging**